File: src/cpp_basic_nodes/src/simple_subscriber.cpp
/*
 * This program does the following
 *      1. Demonstrates how to create a publisher using the NodeHandle
 */

// Include ROS header file
#include <ros/ros.h>
// Include the header file for the message content
#include <std_msgs/String.h>
// Other header files
#include <iostream>

using namespace std;

// Create a callback function
/*
 * This function will be called whenever a message is published on a given topic (with a constant pointer
 * to the message as a parameter)
 */
void subscriber_callback(const std_msgs::String::ConstPtr& msg) {
    // Get the contents of the message
    string st = msg->data;  // Get the data into a string
    ROS_INFO_STREAM("Received a message: '" << st << "'");
}

int main(int argc, char **argv) {
    // Initialize the node
    ros::init(argc, argv, "simple_cpp_subscriber");
    // Initialize the NodeHandle
    ros::NodeHandle nh("~");
    
    // Create a subscriber request
    /*
     * Here, the nodehandle tells the master to route the messages from a topic to this node. A handling function
     * has to deal with the messages. The call returns an object of the Subscriber class
     * First parameter is the topic name (it will be scoped inside the node name)
     * Second parameter is the queue size
     * Third parameter is the callback function
     * Note that the actual callback will happen only in the spin function
     * 
     * Link: http://docs.ros.org/en/latest/api/roscpp/html/classros_1_1NodeHandle.html#a317fe4c05919e0bf3fb5162ccb2f7c28
     * Link (Subscriber class): http://docs.ros.org/en/latest/api/roscpp/html/classros_1_1Subscriber.html
     */
    ros::Subscriber msg_subs = nh.subscribe("subs_str", 1000, subscriber_callback);
    
    // Spin the node (call callbacks when messages are received, etc.)
    ros::spin();
    
    return 0;
}

File: src/py_basic_nodes/scripts/simple_subscriber.py
#!/usr/bin/env python3

"""
    Main objective of this program is to do the following:
        1. Demonstrate how to create a subscriber
"""

# Import ROS Client library
import rospy
# Import system library
import sys
# Import library for the message file
from std_msgs.msg import String

# Create a callback function
"""
    This function will be called with an object to access the message contents as the argument
    every time a message is received on the subscribed topic
"""
def subscriber_callback(msg_obj: String):
    rospy.loginfo("Received a message: '{0}'".format(msg_obj.data))

def main():
    # Initialize the node
    rospy.init_node("simple_py_subscriber", argv=sys.argv)
    
    # Create a subscriber object
    """
        We will use this object to subscribe to topics
        The first argument is topic name
        The second argument is the class (message type)
        The third argument is the 

        Link (Subscriber class): http://docs.ros.org/en/melodic/api/rospy/html/rospy.topics.Subscriber-class.html
        Link (constructor): http://docs.ros.org/en/melodic/api/rospy/html/rospy.topics.Subscriber-class.html#__init__
    """
    sub_obj = rospy.Subscriber("{0}/subs_str".format(rospy.get_name()), String, subscriber_callback, queue_size=1000)
    # Set the node to spin
    rospy.spin()

if __name__ == "__main__":
    try:
        main()
    except rospy.ROSException as e:
        rospy.logfatal("A fatal error has ocurred: {0}".format(e))
